helix-term/src/application.rs

```diff
@@ -951,24 +951,32 @@ impl Application {
             Call::MethodCall(helix_lsp::jsonrpc::MethodCall {
                 method, params, id, ..
             }) => {
-                let call = match MethodCall::parse(&method, params) {
-                    Ok(call) => call,
+                let reply = match MethodCall::parse(&method, params) {
                     Err(helix_lsp::Error::Unhandled) => {
-                        error!("Language Server: Method not found {}", method);
-                        return;
+                        error!(
+                            "Language Server: Method {} not found in request {}",
+                            method, id
+                        );
+                        Err(helix_lsp::jsonrpc::Error {
+                            code: helix_lsp::jsonrpc::ErrorCode::MethodNotFound,
+                            message: format!("Method not found: {}", method),
+                            data: None,
+                        })
```

Comment:
Pyright exits if we send a JSONRPC error for its registerClient request. I'm pretty sure this is a bug in pyright or one of its dependencies.

<details><summary>Log...</summary>

```
2023-02-19T15:46:40.289 helix_lsp::client [INFO] Using custom LSP config: {}
2023-02-19T15:46:40.290 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"general":{"positionEncodings":["utf-32","utf-8","utf-16"]},"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"completion":{"completionItem":{"deprecatedSupport":true,"insertReplaceSupport":true,"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"snippetSupport":false,"tagSupport":{"valueSet":[1]}},"completionItemKind":{}},"hover":{"contentFormat":["markdown"]},"publishDiagnostics":{},"rename":{"dynamicRegistration":false,"honorsChangeAnnotations":false,"prepareSupport":false},"signatureHelp":{"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["markdown"],"parameterInformation":{"labelOffsetSupport":true}}}},"window":{"workDoneProgress":true},"workspace":{"applyEdit":true,"configuration":true,"didChangeConfiguration":{"dynamicRegistration":false},"executeCommand":{"dynamicRegistration":false},"symbol":{"dynamicRegistration":false},"workspaceFolders":true}},"clientInfo":{"name":"helix","version":"22.12 (31b0c758)"},"initializationOptions":{},"processId":102760,"rootPath":"/home/michael/tmp","rootUri":"file:///home/michael/tmp","workspaceFolders":[{"name":"tmp","uri":"file:///home/michael/tmp"}]},"id":0}
2023-02-19T15:46:40.409 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Pyright language server 1.1.294 starting"}}
2023-02-19T15:46:40.409 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Server root directory: /home/michael/tmp/node_modules/pyright/dist/"}}
2023-02-19T15:46:40.409 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Pyright language server 1.1.294 starting" }
2023-02-19T15:46:40.413 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Starting service instance \"tmp\""}}
2023-02-19T15:46:40.413 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":2,"definitionProvider":{"workDoneProgress":true},"declarationProvider":{"workDoneProgress":true},"typeDefinitionProvider":{"workDoneProgress":true},"referencesProvider":{"workDoneProgress":true},"documentSymbolProvider":{"workDoneProgress":true},"workspaceSymbolProvider":{"workDoneProgress":true},"hoverProvider":{"workDoneProgress":true},"documentHighlightProvider":{"workDoneProgress":true},"renameProvider":{"prepareProvider":true,"workDoneProgress":true},"completionProvider":{"triggerCharacters":[".","["],"resolveProvider":true,"workDoneProgress":true,"completionItem":{"labelDetailsSupport":true}},"signatureHelpProvider":{"triggerCharacters":["(",",",")"],"workDoneProgress":true},"codeActionProvider":{"codeActionKinds":["quickfix","source.organizeImports"],"workDoneProgress":true},"executeCommandProvider":{"commands":[],"workDoneProgress":true},"callHierarchyProvider":true}}}
2023-02-19T15:46:40.413 helix_lsp::transport [INFO] <- {"capabilities":{"callHierarchyProvider":true,"codeActionProvider":{"codeActionKinds":["quickfix","source.organizeImports"],"workDoneProgress":true},"completionProvider":{"completionItem":{"labelDetailsSupport":true},"resolveProvider":true,"triggerCharacters":[".","["],"workDoneProgress":true},"declarationProvider":{"workDoneProgress":true},"definitionProvider":{"workDoneProgress":true},"documentHighlightProvider":{"workDoneProgress":true},"documentSymbolProvider":{"workDoneProgress":true},"executeCommandProvider":{"commands":[],"workDoneProgress":true},"hoverProvider":{"workDoneProgress":true},"referencesProvider":{"workDoneProgress":true},"renameProvider":{"prepareProvider":true,"workDoneProgress":true},"signatureHelpProvider":{"triggerCharacters":["(",",",")"],"workDoneProgress":true},"textDocumentSync":2,"typeDefinitionProvider":{"workDoneProgress":true},"workspaceSymbolProvider":{"workDoneProgress":true}}}
2023-02-19T15:46:40.414 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialized","params":{}}
2023-02-19T15:46:40.415 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"method":"client/registerCapability","params":{"registrations":[{"id":"19a6c6b9-cbac-4619-bb83-3fbf546bb213","method":"workspace/didChangeWorkspaceFolders","registerOptions":{}}]}}
2023-02-19T15:46:40.418 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Server root directory: /home/michael/tmp/node_modules/pyright/dist/" }
2023-02-19T15:46:40.418 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Starting service instance \"tmp\"" }
2023-02-19T15:46:40.418 helix_term::application [ERROR] Language Server: Method client/registerCapability not found in request 0
2023-02-19T15:46:40.418 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}
2023-02-19T15:46:40.418 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","error":{"code":-32601,"message":"Method not found: client/registerCapability"},"id":0}
2023-02-19T15:46:40.418 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"python","text":"import json\n\ndef func(a):\n    print(a)\n","uri":"file:///home/michael/tmp/f.py","version":0}}}
2023-02-19T15:46:40.419 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":1,"method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///home/michael/tmp","section":"python"}]}}
2023-02-19T15:46:40.419 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null],"id":1}
2023-02-19T15:46:40.562 helix_lsp::transport [ERROR] err <- "Error: Method not found: client/registerCapability\n"
2023-02-19T15:46:40.562 helix_lsp::transport [ERROR] err <- "    at handleResponse (/home/michael/tmp/node_modules/pyright/dist/pyright-internal/node_modules/vscode-jsonrpc/lib/common/connection.js:565:48)\n"
2023-02-19T15:46:40.562 helix_lsp::transport [ERROR] err <- "    at handleMessage (/home/michael/tmp/node_modules/pyright/dist/pyright-internal/node_modules/vscode-jsonrpc/lib/common/connection.js:345:13)\n"
2023-02-19T15:46:40.562 helix_lsp::transport [ERROR] err <- "    at processMessageQueue (/home/michael/tmp/node_modules/pyright/dist/pyright-internal/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)\n"
2023-02-19T15:46:40.562 helix_lsp::transport [ERROR] err <- "    at Immediate.<anonymous> (/home/michael/tmp/node_modules/pyright/dist/pyright-internal/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)\n"
2023-02-19T15:46:40.562 helix_lsp::transport [ERROR] err <- "    at processImmediate (node:internal/timers:476:21)\n"
2023-02-19T15:46:40.568 helix_lsp::transport [ERROR] err: <- StreamClosed
```
</details>

Reply:
This is actually a big rat's nest in the LSP landscape. There is a bunch of prior discussion about this: https://github.com/microsoft/vscode-languageserver-node/issues/713, https://github.com/microsoft/pyright/issues/1302, https://github.com/neovim/neovim/issues/13448, https://github.com/sublimelsp/LSP/pull/911, https://github.com/neovim/nvim-lspconfig/pull/471

LSP clients either ended up either fully implementing dynamic callback registration or sending dummy responses for registerClient.

The problem is in the https://github.com/microsoft/vscode-languageserver-node dependency which is used by quite a few popular servers. If a server sets up an `onDidChangeWorkspaceFolders` callback, it will try to register `workspace/didChangeWorkspaceFolders`. A change in the way failed Promises are handled in nodejs v15 makes that registration Promise fatal for the server if we send an error.

That dependency sends a registration request for that case because of an ambiguity in the LSP spec: their reasoning is that if workspaceFolders is enabled in the initialization params then that means that the client supports dynamic registration of the workspace folder change notification. workspaceFolders doesn't have a dynamicRegistration setting so you can't explicitly say that you don't support dynamic registration for that capability. I disagree with their interpretation (and it looks like some nvim and sublime folks do too) but it looks unlikely that they will accept a change upstream. This will all be moot if we implement dynamic capability registration anyways so I will add a hack for now like neovim does.

Reply:
Maybe I am missing something but it sounds Luke this is actually solved if the server send the static capabilities correctly https://github.com/microsoft/vscode-languageserver-node/issues/713#issuecomment-754551117

Reply:
I am not sure we even want dynamic capability registration it seems like a pretty large effort with loads of edgecases that provides little advatanges over static registration. Especially since we can opt out of dynamic registration (except in this edgecase). Maybe I am missing some feature this provides but I can't think of any

Reply:
Yeah I think I remember some prior discussion about dynamic capability registration and it not really being desired. Neovim is able to ignore it so I don't think we necessarily need it. I'll remove the TODO for adding it 👍

Reply:
> Maybe I am missing something but it sounds like this is actually solved if the server send the static capabilities correctly [microsoft/vscode-languageserver-node#713 (comment)](https://github.com/microsoft/vscode-languageserver-node/issues/713#issuecomment-754551117)

I believe pyright could fix this by adding a block to https://github.com/microsoft/pyright/blob/d22850b041edbc747ec043553a23bdbff55696d6/packages/pyright-internal/src/languageServerBase.ts#L726-L760

```js
workspace: {
    workspaceFolders: {
        supported: true,
        changeNotifications: false,
    }
}
```

(I'll try it out to confirm.) But they are actually interested in workspace folder change notifications so I think it makes sense for them to set that to `true`. They could gate it on whether the client says that it supports dynamic capability registration but at that point this should be solved in vscode-languageserver-node instead because all other language servers based on vscode-languageserver-node would need to do the same.

Sending `Ok(Null)` for registerClient is still a hack but I'm comfortable with the tradeoff for the sake of compatibility. I think this should definitely be fixed upstream though ultimately.

Reply:
> ```js
> workspace: {
>     workspaceFolders: {
>         supported: true,
>         changeNotifications: false,
>     }
> }
> ```

I actually think the fix would be:

```
workspace: {
    workspaceFolders: {
        supported: true,
        changeNotifications: true,
    }
}
```

If I read https://github.com/microsoft/vscode-languageserver-node/commit/5bc82ac3c1c656aa70615c9e3e065fa04e41506e right it looks like `vscode-languageserver-node` only attemps dynamic registration if static registration has not already occured. So if  `changeNotifications` is `true` they should not send the problematic request

Reply:
Oh good catch I think you're right. Let me give that a try 👍

Reply:
Yep that does it! I'll submit that upstream